### PR TITLE
Shift event handling methods out of UiManager #498

### LINKED ISSUE #498: Shift event handling methods out of UiManager
`UiManager` is catching events such as `JumpToListRequestEvent` which then calls methods to cause the `PersonListPanel` to scroll to the corresponding location.

There is no need for a high-level component such as `UiManager` to catch these events. Rather, the individual components ought to catch these events, as they are already currently handling these events. Using the example given above, the `PersonListPanel` ought to catch the event instead.

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #498 

```
UiManager is catching events such as JumpToListRequestEvent which then 
calls methods to cause the PersonListPanel to scroll to the 
corresponding location.

There is no need for a high-level component such as UiManager to catch 
these events. As such, the individual components that are handling 
these events ought to catch them instead.

Let's shift the catching of events into the classes that are handling 
them.
```

### CODE REVIEW

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 1]
```diff
@@ -219,4 +226,38 @@ private void handleShowHelpEvent(ShowHelpRequestEvent event) {
         logger.info(LogsCenter.getEventHandlingLogMessage(event));
         handleHelp();
     }
+
+    @Subscribe
+    private void handleDataSavingExceptionEvent(DataSavingExceptionEvent event) {
+        logger.info(LogsCenter.getEventHandlingLogMessage(event));
+        showFileOperationAlertAndWait("Could not save data", "Could not save data to file", event.exception);
```
Comment (yamgent) [inline src/main/java/seedu/address/ui/MainWindow.java]: This brings us to a long chain of method calls. Is there any way we can cut out the middle-man methods?

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 2]
```diff
@@ -72,5 +83,4 @@ protected void updateItem(ReadOnlyPerson person, boolean empty) {
             }
         }
     }
-
```
Comment (yamgent) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: Unnecessary line change

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/5] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/fce686f81ee76f2dd5251a433010b9e3928dedee)
* [v1 2/5] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/03b85c9f2d15b456b344c92468a74a14617f8be8)
* [v1 3/5] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/573bbd1e8a43face9a88e9238bdba1c15701ecf7)
* [v1 4/5] [Move UiManager\#handleDataSavingExceptionEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/cc51552528cf2fcc99703b77f33971f46485c04a)
* [v1 5/5] [Inline MainWindow\#handleDataSavingExceptionEvent helper methods](https://github.com/se-edu/addressbook-level4/pull/499/commits/076b5b48339886e6dd1f3b9ec082d902c24827af)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...076b5b48339886e6dd1f3b9ec082d902c24827af))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/5] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/fce686f81ee76f2dd5251a433010b9e3928dedee)
* [v2 2/5] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/03b85c9f2d15b456b344c92468a74a14617f8be8)
* [v2 3/5] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/573bbd1e8a43face9a88e9238bdba1c15701ecf7)
* [v2 4/5] [Move UiManager\#handleDataSavingExceptionEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/dc1e589ff8178a9ae97cc1c179da682369b0d058)
* [v2 5/5] [Inline MainWindow\#handleDataSavingExceptionEvent helper methods](https://github.com/se-edu/addressbook-level4/pull/499/commits/cb7f4ad1ea1244deb00f4ae53411b7887350ddc1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...cb7f4ad1ea1244deb00f4ae53411b7887350ddc1)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/7f9b0f491d1f9ce295367c56ff9329553d773a2c~1...7f9b0f491d1f9ce295367c56ff9329553d773a2c))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/5] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/0d70d245ef1808daf63ebefa717e2decf578096d)
* [v3 2/5] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/fea64079bed18d7a7d047c54112cee5136d52487)
* [v3 3/5] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/29d274486a3ba19b65ee5f2c39290a42160f7cf0)
* [v3 4/5] [Move UiManager\#handleDataSavingExceptionEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/96aac6731445bdce827d8629dcdebd4e5da41c72)
* [v3 5/5] [Inline MainWindow\#handleDataSavingExceptionEvent helper methods](https://github.com/se-edu/addressbook-level4/pull/499/commits/3679d3014f44dbb5feabc5019eafab77a1a939af)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...3679d3014f44dbb5feabc5019eafab77a1a939af)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/7d5b2b4ee27a6e6f2ff635cef8b93c2ccd2f41c2~1...7d5b2b4ee27a6e6f2ff635cef8b93c2ccd2f41c2))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@yamgent made the changes, also check the inter-diff as I have updated some method calls

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/6] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/0d70d245ef1808daf63ebefa717e2decf578096d)
* [v4 2/6] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/fea64079bed18d7a7d047c54112cee5136d52487)
* [v4 3/6] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/29d274486a3ba19b65ee5f2c39290a42160f7cf0)
* [v4 4/6] [Move UiManager\#handleDataSavingExceptionEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/96aac6731445bdce827d8629dcdebd4e5da41c72)
* [v4 5/6] [Inline MainWindow\#handleDataSavingExceptionEvent helper methods](https://github.com/se-edu/addressbook-level4/pull/499/commits/3679d3014f44dbb5feabc5019eafab77a1a939af)
* [v4 6/6] [MainWindow: Remove unused method](https://github.com/se-edu/addressbook-level4/pull/499/commits/3caaccf7108134a6374bd9102236bcfaa0686b90)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...3caaccf7108134a6374bd9102236bcfaa0686b90)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/9096348178e9984495855305ff4a1884e2dcd99f~1...9096348178e9984495855305ff4a1884e2dcd99f))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
Did an additional commit (commit 6).

File: src/main/java/seedu/address/ui/BrowserPanel.java [hunk 3]
```diff
@@ -33,9 +40,10 @@ public BrowserPanel(Pane placeholder) {
         placeholder.getChildren().add(browser);
 
         loadDefaultPage();
+        registerAsAnEventHandler(this);
     }
 
-    public void loadPersonPage(ReadOnlyPerson person) {
+    private void loadPersonPage(ReadOnlyPerson person) {
```
Comment (limmlingg) [inline src/main/java/seedu/address/ui/BrowserPanel.java]: Same as before, I don't think this is relevant to this commit.

File: src/main/java/seedu/address/ui/PersonListPanel.java [hunk 4]
```diff
@@ -51,13 +55,19 @@ private void setEventHandlerForSelectionChangeEvent() {
                 });
     }
 
-    public void scrollTo(int index) {
+    private void scrollTo(int index) {
```
Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: @yamgent what do you think? imo this change makes sense because in this commit `MainWindow` is no longer calling `PersonListPanel#scrollTo(int)` (and there's no one else calling this method as well), thus we can make it private. 

Well I can create another commit for this line too.... but seems unnecessary.

**Comment by Zhiyuan-Amos** [comment]
@limmlingg [6/6] is in the scope of this PR, just that perhaps I didn't draw the link clearly.

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/6] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/72a4d6e9e2b164dea4559a67470ebef17155bb33)
* [v5 2/6] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/bae98b7680c40481c57866e187f0d8cea4e8f7d4)
* [v5 3/6] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/29ebc3a2f9b22dbf1c9908fa4d991beefffed7c4)
* [v5 4/6] [Move UiManager\#handleDataSavingExceptionEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/0cb82d3ae62cf63eac1c631b5e7af1baae3f2c15)
* [v5 5/6] [MainWindow: Remove MainWindow\#getPersonListPanel\(\)](https://github.com/se-edu/addressbook-level4/pull/499/commits/fa6d1c8d00c8cc6ef956576d253bd5dfdcdefce6)
* [v5 6/6] [Inline MainWindow\#handleDataSavingExceptionEvent helper methods](https://github.com/se-edu/addressbook-level4/pull/499/commits/e5a59f39c9ac9596bac27cd004ab55995f9dd4b3)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...e5a59f39c9ac9596bac27cd004ab55995f9dd4b3)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/650c139e8c5dc72598699762ae0344ddb0521de0~1...650c139e8c5dc72598699762ae0344ddb0521de0))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/ui/PersonListPanel.java (hunk 4 above)
Comment (yamgent) [inline src/main/java/seedu/address/ui/PersonListPanel.java]: @Zhiyuan-Amos I think you should put it in the separate commit, because otherwise the commit is doing too much things.

**Comment by CanIHasReview** [comment]
# v6

@Zhiyuan-Amos submitted v6 for review.

* [v6 1/8] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/72a4d6e9e2b164dea4559a67470ebef17155bb33)
* [v6 2/8] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/50120f8d03f3cf7b40b1002f90383accf2ffa1cb)
* [v6 3/8] [Update PersonListPanel\#scrollTo\(int\) visibility modifier to private](https://github.com/se-edu/addressbook-level4/pull/499/commits/c42e09845797c3b6c05171107c4a77a3141ef425)
* [v6 4/8] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/ff9ec1a586f9e1f3259d85c882555004d6a50d0b)
* [v6 5/8] [Update BrowserPanel\#loadPersonPage\(…\) visibility modifier to private](https://github.com/se-edu/addressbook-level4/pull/499/commits/a998aa367b362e051da867427959b0388fc0f9e0)
* [v6 6/8] [Move UiManager\#handleDataSavingExceptionEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/ab9f8d4a3f8af1275d69072965db933de5b92a1d)
* [v6 7/8] [MainWindow: Remove MainWindow\#getPersonListPanel\(\)](https://github.com/se-edu/addressbook-level4/pull/499/commits/637f2f6029e0286e52c6ea609ac438852f351217)
* [v6 8/8] [Inline MainWindow\#handleDataSavingExceptionEvent helper methods](https://github.com/se-edu/addressbook-level4/pull/499/commits/23f3ee781be516a9c87254330530bfb80135c630)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...23f3ee781be516a9c87254330530bfb80135c630)) ([:chart_with_upwards_trend: Interdiff between v5 and v6](https://github.com/se-edu/addressbook-level4/compare/04326efeffb9350f1fe909dc4f2018f614669ff7~1...04326efeffb9350f1fe909dc4f2018f614669ff7))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/6/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@limmlingg

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 5]
```diff
@@ -197,16 +212,38 @@ private void handleExit() {
...
+    @Subscribe
+    private void handleShowHelpEvent(ShowHelpRequestEvent event) {
+        logger.info(LogsCenter.getEventHandlingLogMessage(event));
+        handleHelp();
     }
 
-    void releaseResources() {
-        browserPanel.freeResources();
+    @Subscribe
+    private void handleDataSavingExceptionEvent(DataSavingExceptionEvent event) {
```
Comment (pyokagan) [inline src/main/java/seedu/address/ui/MainWindow.java]: "data saving exception" doesn't seem specific to the "main window", and would probably be better served remaining in UiManager since it is a global UI thing.

Comment (Zhiyuan-Amos) [inline src/main/java/seedu/address/ui/MainWindow.java]: Yes, but I would argue that different windows may want to handle this exception differently. Suppose we have a `MainWindowTwo`, which is another parent View Controller (like `MainWindow`). `MainWindowTwo` may want to handle this exception differently instead of showing an `AlertDialog`, as such I think it seems reasonable to allow `MainWindow` to handle this exception.

Or perhaps, if we are on `MainWindowTwo`, no `DataSavingExceptionEvent` should be raised because there's no saving involved, but because there's some bug in the code, somehow this event is raised, causing UiManager to catch it and handle it.

Though I get your argument: if I'm reading you rightly, you're saying that even if we have `MainWindowTwo` etc, they are likely to use the same way of handling it too, so we should leave it in the UiManager :P

**Comment by Zhiyuan-Amos** [comment]
Made the changes that Paul requested :>

**Comment by CanIHasReview** [comment]
# v7

@Zhiyuan-Amos submitted v7 for review.

* [v7 1/5] [Move UiManager\#handleShowHelpEvent\(…\) to MainWindow](https://github.com/se-edu/addressbook-level4/pull/499/commits/2a4b7aa165fc368eb95ddc87e3e374530cf26902)
* [v7 2/5] [Move UiManager\#handleJumpToListRequestEvent\(…\) to PersonListPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/481e58e9256c2ca5411d85e17d75c023d6316781)
* [v7 3/5] [Update PersonListPanel\#scrollTo\(int\) visibility modifier to private](https://github.com/se-edu/addressbook-level4/pull/499/commits/15cf4b14b80b91fc6d2259105cd88411849dbc33)
* [v7 4/5] [Move UiManager\#handlePersonPanelSelectionChangedEvent\(…\) to BrowserPanel](https://github.com/se-edu/addressbook-level4/pull/499/commits/d725516f91920b256dcb7966b8b78ff37bb770f0)
* [v7 5/5] [Update BrowserPanel\#loadPersonPage\(…\) visibility modifier to private](https://github.com/se-edu/addressbook-level4/pull/499/commits/be4b7fbe305992b86a4d3ab63bee6f65cbb7f770)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/89889b9a0bb1c75b945d41466adc4bec50f9a72e...be4b7fbe305992b86a4d3ab63bee6f65cbb7f770)) ([:chart_with_upwards_trend: Interdiff between v6 and v7](https://github.com/se-edu/addressbook-level4/compare/712710baacb26582daa1c31dee42f28bed287a0e~1...712710baacb26582daa1c31dee42f28bed287a0e))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/499/7/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@damithc I can remove commit 1 if you think that it shouldn't be there :)
